src/main/haskell/kore/src/Kore/Exec.hs

```diff
@@ -0,0 +1,231 @@
+import           Kore.Step.OrOfExpandedPattern
+                 ( OrOfExpandedPattern )
+import qualified Kore.Step.OrOfExpandedPattern as OrOfExpandedPattern
+import qualified Kore.Step.PredicateSubstitution as PredicateSubstitution
+                 ( toPredicate )
+import           Kore.Step.Search
+                 ( SearchType (..), search )
+import qualified Kore.Step.Search as Search
+import           Kore.Step.Simplification.Data
+                 ( PredicateSubstitutionSimplifier (..),
+                 PureMLPatternSimplifier, SimplificationProof (..),
+                 Simplifier )
+import qualified Kore.Step.Simplification.ExpandedPattern as ExpandedPattern
+import qualified Kore.Step.Simplification.PredicateSubstitution as PredicateSubstitution
+import qualified Kore.Step.Simplification.Simplifier as Simplifier
+                 ( create )
+import           Kore.Step.Step
+import           Kore.Step.StepperAttributes
+                 ( StepperAttributes (..) )
+import           Kore.Substitution.Class
+                 ( Hashable, substitute )
+import qualified Kore.Substitution.List as ListSubstitution
+import           Kore.Variables.Fresh
+                 ( FreshVariable )
+
+data KoreSearchParameters = KoreSearchParameters
+    { searchPattern :: !(ExpandedPattern.ExpandedPattern Object Variable)
+    -- ^ Pattern to match during execution
+    , boundLimit :: !(Limit Natural)
+    -- ^ The maximum bound on the number of search matches
+    , searchType :: !SearchType
+    -- ^ The type of search to perform
+    }
+
+exec ::
+    KoreIndexedModule StepperAttributes
+    -> CommonPurePattern Object
+    -> Limit Natural
+    -> ([AxiomPattern Object] -> Strategy (Prim (AxiomPattern Object)))
+    -> Maybe KoreSearchParameters
+    -> Simplifier (CommonPurePattern Object)
+exec indexedModule purePattern stepLimit strategy searchConfig = do
+    let
+        metadataTools = extractMetadataTools indexedModule
+        MetadataTools { symbolOrAliasSorts } = metadataTools
+    axiomsAndSimplifiers <-
+        makeAxiomsAndSimplifiers indexedModule metadataTools
+    let
+        (rewriteAxioms, simplifier, substitutionSimplifier) =
+            axiomsAndSimplifiers
+        runStrategy' pat =
+            runStrategy
+                (transitionRule metadataTools substitutionSimplifier simplifier)
+                (Limit.replicate stepLimit (strategy rewriteAxioms))
+                (pat, mempty)
+        expandedPattern = makeExpandedPattern purePattern
+    simplifiedPatterns <-
+        ExpandedPattern.simplify
+            metadataTools substitutionSimplifier simplifier expandedPattern
+    let
+        initialPattern =
+            case OrOfExpandedPattern.extractPatterns (fst simplifiedPatterns) of
+                [] -> ExpandedPattern.bottom
+                (config : _) -> config
+    executionTree <- runStrategy' initialPattern
+    case searchConfig of
```

Comment:
I would probably let there be two functions, `exec` (concrete execution) and `search` (search, i.e. symbolic execution).